Use REST API instead of XmlHttpRequest

dregad opened the pull request:
XHR was deprecated in Mantis 2.3.0.

Publish REST API endpoints to replace the corresponding XHR functions:

- xmlhttprequest_plugin_announce_dismiss
  => ./plugins/Announce/dismiss/{context_id}
- xmlhttprequest_plugin_announce_add_context
  => ./plugins/Announce/context/{message_id}

Fixes #21

vboctor commented on Announce.php:
```diff
@@ -103,4 +105,82 @@ function schema() {
+		$t_user_id = auth_get_current_user_id();
+
+		# Set the reference Context Id for dismissal
+		$t_context_id = (int)$args['context_id'];
```
Check that the context_id field was passed.

dregad replied:
Slim does that for me, AFAICT - the following request http://path.to/api/rest/plugins/Announce/dismiss (without providing a *context_id*) yields a 404

vboctor replied:
Make sure to test this with `http://path.to/api/rest/plugins/Announce/dismiss/`.

dregad replied:
> Make sure to test this with http://path.to/api/rest/plugins/Announce/dismiss/

I did -
```console
$ curl -si -X POST http://localhost/mantis/api/rest/plugins/Announce/dismiss |head -1
HTTP/1.1 404 Not Found
$ curl -si -X POST http://localhost/mantis/api/rest/plugins/Announce/dismiss/ |head -1
HTTP/1.1 404 Not Found
```

> Also you may want to validate the input. For example, is the context_id > 0?

This is actually done by *AnnounceContext::load_by_id()*, 

```console
$ curl -si -X POST http://localhost/mantis/api/rest/plugins/Announce/dismiss/0 |head -1
HTTP/1.1 400 Bad Request
```

vboctor commented on Announce.php:
```diff
@@ -103,4 +105,82 @@ function schema() {
+			return $response->withStatus( HTTP_STATUS_BAD_REQUEST );
+		}
+
+		plugin_push_current( $this->basename );
```
I suggest having push as the first statement and pop as the last before return, similar to what you had at `xmlhttprequest_plugin_announce_dismiss()`.  This avoids confusion in the future.  Hopefully we will remove the need for these at one point.

dregad replied:
Since pushing is only needed when calling plugin API functions and none are used before that point, the idea was to avoid having to call *plugin_pop_current()* before the return statement. I'll gladly revisit this if and when it is no longer necessary to push/pop.

vboctor commented on Announce.php:
```diff
@@ -103,4 +105,82 @@ function schema() {
+		$t_app->group(
+			plugin_route_group(),
+			function() use ( $t_app, $t_plugin ) {
+				$t_app->get( '/dismiss/{context_id}', [$t_plugin, 'route_dismiss'] );
```
This seems like a `POST` rather than a `GET`.

dregad replied:
Right.

dregad commented:
Thanks for your review. I changed to a POST route, and refactored the code to return different status codes but no actual content. Let me know what you think about this approach.

vboctor commented on Announce.php:
```diff
@@ -103,4 +105,87 @@ function schema() {
+		if( db_query( $t_query, $t_param ) ) {
+			$t_status = HTTP_STATUS_SUCCESS;
+		} else {
+			$t_status = HTTP_STATUS_INTERNAL_SERVER_ERROR;
```
It is generally helpful when you return failure status code like bad request, internal server error, etc to also return a string explaining the error.  So it is easy to understand what is going on from the client side.

dregad replied:
Thanks for the advice, will update the code accordingly.
Should the error message be localized, or is it OK to just use English ?

vboctor replied:
English only - no localization for status code errors.

dregad commented:
Rebased and refactored the 2nd AJAX call (add context) to use REST also; fixed a PHP notice and removed deprecated synchronous AJAX call.

dregad commented:
@vboctor your feedback on the latest changes (new REST API endpoint to _add context_) would be appreciated

vboctor commented on Announce.php:
```diff
@@ -103,4 +105,143 @@ function schema() {
+			plugin_route_group(),
+			function() use ( $t_app, $t_plugin ) {
+				$t_app->post( '/dismiss/{context_id}', [$t_plugin, 'route_dismiss'] );
+				$t_app->get( '/context/{message_id}', [$t_plugin, 'route_new_context'] );
```
The name of the method indicates that you are creating a context rather than getting info/form for creating a new context.  I would rename the method.

dregad replied:
Makes sense, will do

vboctor commented on Announce.php:
```diff
@@ -103,4 +105,143 @@ function schema() {
+	 * @param array $args
+	 * @return Psr\Http\Message\ResponseInterface
+	 */
+	public function route_new_context( $request, $response, $args) {
```
Is there a reason the REST API returns actual HTML rather than json info that is needed to render the HTML form?

dregad replied:
No particular reason, other than reusing the code from the legacy xmlhttprequest endpoint. Are you suggesting to have the HTML generated in the javascript instead ?

dregad replied:
Note: the advantage of generating the HTML on the PHP side is that the code can leverage the MantisBT print APIs for option lists. To return raw data in JSON format implies code duplication, so I don't really see the added value.

vboctor replied:
OK, I haven't got to define patterns around returning UI vs. server side rendering.  You can do whatever makes sense for now.  We can always re-visit if it makes sense later.  Just moving to REST API is a good step.

vboctor commented on files/list_action.js:
```diff
@@ -53,9 +53,8 @@ jQuery(document).ready(function($) {
-			async: false,
 			dataType: "html",
-			url: "xmlhttprequest.php?entrypoint=plugin_announce_add_context&row=1&message_id="+message_id,
+            url: Announce.rest_api('context/') + message_id,
```
Indentation seem to be broken.

dregad replied:
Indeed, will fix.

vboctor commented on files/announce.js:
```diff
@@ -44,17 +60,12 @@ jQuery(document).ready(function($) {
-					)
-				}
+			type: 'POST',
+            // Using the full URL (through index.php) to avoid issues on sites
```
Indentation seem to be broken.

dregad replied:
Will fix.
